### PR TITLE
Fix issue where FAR results were not getting sent until they were fully complete

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Accidentally regressed this with some refactorings in this area.

### CODE REVIEW

File: src/Workspaces/Remote/ServiceHub/Services/SymbolFinder/RemoteSymbolFinderService.cs [hunk 1]
```diff
@@ -227,19 +227,14 @@ public ValueTask OnDefinitionFoundAsync(SymbolGroup group, CancellationToken can
...
-                {
-                    result.Add((
-                        SerializableSymbolGroup.Dehydrate(_solution, group, cancellationToken),
-                         SerializableSymbolAndProjectId.Dehydrate(_solution, symbol, cancellationToken),
-                         SerializableReferenceLocation.Dehydrate(location, cancellationToken)));
-                }
+                var dehydrated = references.SelectAsArray(
+                    r => (SerializableSymbolGroup.Dehydrate(_solution, r.group, cancellationToken),
+                          SerializableSymbolAndProjectId.Dehydrate(_solution, r.symbol, cancellationToken),
+                          SerializableReferenceLocation.Dehydrate(r.location, cancellationToken)));
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Services/SymbolFinder/RemoteSymbolFinderService.cs:236]: i'm not sure how i missed this.  this is the server-oop sending point.  When i switched this to IAsyncEnumerable, it meant that this was pulling on the entire stream, waiting for it to finish, before sending anything over to the client.  

switched to the ImmutableArray auto-chunking form. 

The semantics here are that as the server is sending this IA over, the ProducerConumser is still searching, and buffering up the results found after the last chunk was send to 'consumeItems'.  After this call to the client finishes, we'll get another consumeItems call with that buffer's content.  And that repeats until the producer is done.

**Comment by CyrusNajmabadi** [comment]
Have validated this fixes hte issue.

File: src/Workspaces/Core/Portable/FindSymbols/SymbolFinder.FindReferencesServerCallback.cs [hunk 2]
```diff
@@ -65,39 +65,37 @@ public async ValueTask OnDefinitionFoundAsync(SerializableSymbolGroup dehydrated
             await progress.OnDefinitionFoundAsync(symbolGroup, cancellationToken).ConfigureAwait(false);
         }
 
-        public ValueTask OnReferencesFoundAsync(
+        public async ValueTask OnReferencesFoundAsync(
             ImmutableArray<(SerializableSymbolGroup serializableSymbolGroup, SerializableSymbolAndProjectId serializableSymbol, SerializableReferenceLocation reference)> references,
             CancellationToken cancellationToken)
         {
-            return progress.OnReferencesFoundAsync(ConvertAsync(cancellationToken), cancellationToken);
+            using var _ = ArrayBuilder<(SymbolGroup group, ISymbol symbol, ReferenceLocation location)>.GetInstance(references.Length, out var rehydrated);
```
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolFinder.FindReferencesServerCallback.cs:72]: >ArrayBuilder [](http://example.com/codeflow?start=26&length=12)

What's the point of using an arraybuilder here, won't there only ever be a single item added to it before it gets cleared?

File: src/Workspaces/Remote/ServiceHub/Services/SymbolFinder/RemoteSymbolFinderService.cs (hunk 1 above)
Comment (ToddGrun) [reply src/Workspaces/Remote/ServiceHub/Services/SymbolFinder/RemoteSymbolFinderService.cs:236]: I've noticed a big lag on getting the initial reference the last couple days, just thought my machine was bugging out on me

Comment (CyrusNajmabadi) [reply src/Workspaces/Remote/ServiceHub/Services/SymbolFinder/RemoteSymbolFinderService.cs:236]: :'(

File: src/Workspaces/Core/Portable/FindSymbols/SymbolFinder.FindReferencesServerCallback.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/FindSymbols/SymbolFinder.FindReferencesServerCallback.cs:72]: fixed!

File: src/Workspaces/Remote/ServiceHub/Services/SymbolFinder/RemoteSymbolFinderService.cs [hunk 3]
```diff
@@ -227,19 +227,14 @@ public ValueTask OnDefinitionFoundAsync(SymbolGroup group, CancellationToken can
...
-                using var _ = ArrayBuilder<(SerializableSymbolGroup, SerializableSymbolAndProjectId, SerializableReferenceLocation)>.GetInstance(out var result);
-                await foreach (var (group, symbol, location) in references)
-                {
-                    result.Add((
-                        SerializableSymbolGroup.Dehydrate(_solution, group, cancellationToken),
-                         SerializableSymbolAndProjectId.Dehydrate(_solution, symbol, cancellationToken),
-                         SerializableReferenceLocation.Dehydrate(location, cancellationToken)));
-                }
+                var dehydrated = references.SelectAsArray(
+                    r => (SerializableSymbolGroup.Dehydrate(_solution, r.group, cancellationToken),
```
Comment (ToddGrun) [inline src/Workspaces/Remote/ServiceHub/Services/SymbolFinder/RemoteSymbolFinderService.cs:234]: >r  [](http://example.com/codeflow?start=20&length=2)

make static?

Comment (ToddGrun) [reply src/Workspaces/Remote/ServiceHub/Services/SymbolFinder/RemoteSymbolFinderService.cs:234]: Just noticed that the ArrayBuilder SelectAsArray methods should be using a FixedSizeBuilder. I can do that tomorrow , but it's sadly going to require compiler reviews.

**Review by ToddGrun — APPROVED** [review]
:shipit:

File: src/Workspaces/Remote/ServiceHub/Services/SymbolFinder/RemoteSymbolFinderService.cs (hunk 3 above)
Comment (ToddGrun) [reply src/Workspaces/Remote/ServiceHub/Services/SymbolFinder/RemoteSymbolFinderService.cs:234]: oh, woof, FixedSizeArrayBuilder isn't defined at the compiler level. never mind.

**Pull request merged**